### PR TITLE
apply Catalog API Blocks and Limit data type from Double to BigDecimal

### PR DESCRIPTION
**Pull request by xsalefter**
based on comment on this issue: https://github.com/killbill/killbill/issues/1469#issuecomment-1028580177 .

Related Pull Request in API: https://github.com/killbill/killbill-api/pull/84

### CODE REVIEW

File: catalog/src/main/java/org/killbill/billing/catalog/DefaultLimit.java [hunk 1]
```diff
@@ -40,9 +41,19 @@
     @XmlIDREF
     private DefaultUnit unit;
 
+    /**
+     * FIXME-1469 Catalog: This is need to be converted to BigDecimal. However, if we do, then TestXmlReader.testCatalogLoad()
+     *   will fails. It is because in some XML (test file), max and or min value is not set, and lead to null value.
```
Comment (sbrossie) [inline catalog/src/main/java/org/killbill/billing/catalog/DefaultLimit.java:46]: We need to change this field and come up with a fix for the test to pass - perhaps as simple as updating the xml fail with a default value ?

File: catalog/src/main/java/org/killbill/billing/catalog/DefaultPrice.java [hunk 2]
```diff
@@ -103,7 +103,7 @@ public ValidationErrors validate(final StandaloneCatalog catalog, final Validati
 
     @Override
     public void initialize(final StandaloneCatalog catalog) {
-        super.initialize(catalog);
+        // super.initialize(catalog);
```
Comment (sbrossie) [inline catalog/src/main/java/org/killbill/billing/catalog/DefaultPrice.java:106]: Why are we removing this `initialize` call?

File: catalog/src/main/java/org/killbill/billing/catalog/DefaultTieredBlock.java [hunk 3]
```diff
@@ -83,27 +84,27 @@ public boolean equals(final Object o) {
...
-        return Double.compare(that.max, max) == 0;
+        return max.compareTo(that.max) == 0;
     }
 
     @Override
     public int hashCode() {
         int result = super.hashCode();
         final long temp;
-        temp = Double.doubleToLongBits(max);
+        temp = Double.doubleToLongBits(max.doubleValue());
```
Comment (sbrossie) [inline catalog/src/main/java/org/killbill/billing/catalog/DefaultTieredBlock.java:94]: No more `double`, needs to be changed.

File: catalog/src/main/java/org/killbill/billing/catalog/dao/CatalogOverrideBlockDefinitionModelDao.java [hunk 4]
```diff
@@ -28,8 +28,8 @@
     private String currency;
     private BigDecimal price;
     private String blockType;
-    private double bsize;
-    private double max;
+    private double bsize; // FIXME-1469 Catalog: Should we change this as well
```
Comment (sbrossie) [inline catalog/src/main/java/org/killbill/billing/catalog/dao/CatalogOverrideBlockDefinitionModelDao.java:31]: Yes, should be changed.

File: invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalCapacityUsageInArrear.java [hunk 5]
```diff
@@ -126,7 +126,8 @@ UsageCapacityInArrearAggregate computeToBeBilledCapacityInArrear(final List<Roll
                 final Limit tierLimit = getTierLimit(cur, ro.getUnitType());
                 // We ignore the min and only look at the max Limit as the tiers should be contiguous.
                 // Specifying a -1 value for last max tier will make the validation works
-                if (tierLimit.getMax() != (double) -1 && ro.getAmount().doubleValue() > tierLimit.getMax()) {
+                if (tierLimit.getMax().compareTo(new BigDecimal("-1")) != 0 && ro.getAmount().compareTo(tierLimit.getMax())> 0) {
+                // if (tierLimit.getMax() != (double) -1 && ro.getAmount().doubleValue() > tierLimit.getMax()) {
```
Comment (sbrossie) [inline invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalCapacityUsageInArrear.java:130]: Extra comment?

**Review by sbrossie — CHANGES_REQUESTED** [review]
Left a couple of comments to address, and also running the test locally, I get a test failure: ```/testing-mysql-server17086277404870518227/mysql.sock databaseb008b64d
2022-02-08T19:14:26.581+0000 [main] ERROR org.killbill.billing.GuicyKillbillTestSuiteWithEmbeddedDB - **********************************************************************************************
[ERROR] Tests run: 111, Failures: 1, Errors: 0, Skipped: 40, Time elapsed: 17.941 s <<< FAILURE! - in TestSuite
[ERROR] org.killbill.billing.catalog.dao.TestCatalogOverrideDao.testGetOverriddenPlanPhasesWithUsageOverrides  Time elapsed: 0.056 s  <<< FAILURE!
java.lang.AssertionError: expected [1] but found [1.0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:617)
	at org.killbill.billing.catalog.dao.TestCatalogOverrideDao.testGetOverriddenPlanPhasesWithUsageOverrides(TestCatalogOverrideDao.java:152)
```

**Comment by sbrossie** [comment]
@xsalefter I just merged your `killbill-api` PR, and as a result a snapshot was created. Could you update the value of  [killbill-api.version](https://github.com/xsalefter/killbill/blob/fix_1469-catalog_api/pom.xml#L63) to `0.54.0-0010563-SNAPSHOT` - this will allow `ci` jobs to work once your code is in place?

File: catalog/src/main/resources/org/killbill/billing/catalog/migration/V20220208000000__change_usage_and_catalog_attrs_type.sql [hunk 6]
```diff
@@ -0,0 +1,4 @@
+alter table rolled_up_usage modify amount decimal(17, 9) not null;
```
[the remark this answers is not in the document]
Comment (xsalefter) [reply catalog/src/main/resources/org/killbill/billing/catalog/migration/V20220208000000__change_usage_and_catalog_attrs_type.sql]: This one was typo hahahah.. anyway, fixed.

**Pull request merged**